### PR TITLE
H284: EP15+Sobol-anti-K=3+6-res+mirror — triple-mechanism SOTA compound

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**Combine Sobol QMC direction sampling with anti-thetic pairing in the full 6-res×mirror stack, using your H275 EP15 checkpoint — triple-mechanism compound.**

You just landed H275 as SOTA: EP15 + anti-K=3 + σ=5e-4 + 6-res+mirror → **val 5.9243 / test 5.7690 / test_WSS 6.6743**. Your H275 used i.i.d. Gaussian directions for the K=3 anti-thetic pairs. The question: does Sobol QMC direction sampling (+~0.5bp/channel, Finding UU) stack with anti-thetic linear-term cancellation (Finding VV/ZZ) at EP15?

Key findings in play:
- **Finding UU (frieren H271)**: Sobol K=5 QMC beats random K=5 by −0.5bp/channel at EP13; test_WSS −5.0bp
- **Finding VV (fern H274)**: Anti-thetic K=3 zeros the linear Taylor term ∇f(w)·ε at EP13
- **Finding ZZ (your H275)**: EP15 × anti-K=3 is super-additive (+0.28bp val, +0.51bp test bonus vs pure-additive)

**Predicted val_abupt: 5.921–5.924** (−0–5bp vs H275 5.9243). Same compute as H275 (~240 min on DDP×8).

### Step 1: Patch eval_tta_h252.py to support Sobol direction sampling

`eval_tta_h252.py` (your H275 script) already supports anti-thetic noise. `eval_h271_sobol_tta.py` supports Sobol. You need to add Sobol support to h252. Minimal patch:

**A. Add config fields to `EvalConfig` dataclass (after `antithetic_noise: bool = False`):**
```python
weight_noise_mode: str = "random"   # "random" | "sobol"
sobol_proj_dim: int = 1024
sobol_seed: int = 43
sobol_B_seed: int = 42
sobol_chunk_size: int = 65536
```

**B. Copy Sobol utility functions verbatim from `eval_h271_sobol_tta.py`:**
- `_draw_sobol_std_normal` (lines ~486–500)
- `_sobol_flat_eps` (lines ~506–535)

These are standalone utilities with no dependencies on the h271 global state.

**C. In the noise-draw path inside `eval_tta_h252.py`**, replace the Gaussian draw:
```python
# Before:
eps = torch.randn_like(p_flat) * sigma
# After:
if weight_noise_mode == "sobol":
    eps = _sobol_flat_eps(
        state_dict=model.state_dict(),
        sigma=sigma,
        proj_dim=sobol_proj_dim,
        sobol_seed=sobol_seed + k_index,   # different seed per pair
    )
else:
    eps = torch.randn_like(p_flat) * sigma
```

The anti-thetic logic (apply +eps then −eps per pair) stays unchanged.

**D. Add CLI flags:**
```python
parser.add_argument("--weight-noise-mode", default="random", choices=["random", "sobol"])
parser.add_argument("--sobol-proj-dim", type=int, default=1024)
parser.add_argument("--sobol-seed", type=int, default=43)
```

### Step 2: Verify EP15 checkpoint

```bash
H185_EP15_CKPT="outputs/drivaerml/run-0gjfv45i/checkpoint_ep15.pt"
ls -lh $H185_EP15_CKPT
```

### Step 3: Smoke-check (< 5 min)

```bash
torchrun --standalone --nproc-per-node=8 eval_tta_h252.py \
  --checkpoint $H185_EP15_CKPT \
  --resolutions "32768" \
  --eval-modes "weight_noise_mirror_res_avg" \
  --weight-noise-sigma 5e-4 \
  --weight-noise-passes 1 \
  --antithetic-noise \
  --weight-noise-mode sobol \
  --sobol-proj-dim 1024 \
  --batch-size 4 --num-workers 2 \
  --agent edward --wandb-group h284-smoke
```

### Step 4: Full run (after smoke passes)

```bash
torchrun --standalone --nproc-per-node=8 eval_tta_h252.py \
  --checkpoint $H185_EP15_CKPT \
  --resolutions "32768,49152,65536,81920,98304,131072" \
  --eval-modes "weight_noise_mirror_res_avg" \
  --weight-noise-sigma 5e-4 \
  --weight-noise-passes 3 \
  --antithetic-noise \
  --weight-noise-mode sobol \
  --sobol-proj-dim 1024 \
  --batch-size 2 --num-workers 4 \
  --agent edward \
  --wandb-group "h284-edward-sobol-anti-K3-ep15-stack" \
  --wandb-name "edward/h284-ep15-sobol-anti-K3-stack"
```

NOTE — confirmed flag names from eval_tta_h252.py source:
- `--weight-noise-passes 3` (NOT `--n-weight-noise-passes`)
- `--antithetic-noise` (NOT `--antithetic`)
- `--eval-modes "weight_noise_mirror_res_avg"` (exact mode string)

`--weight-noise-passes 3` + `--antithetic-noise` = **K=3 PAIRS = 6 forward passes** per config. Sobol changes direction-sampling distribution only, not count.

### Step 5: Post terminal SENPAI-RESULT

```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<run-id>"],"primary_metric":{"name":"val_abupt_ep15_sobol_anti_K3_stack","value":<val>},"test_metric":{"name":"test_abupt_ep15_sobol_anti_K3_stack","value":<test>}}
```

Include the triple-mechanism comparison table:

| Recipe | mechanism | val_abupt | test_abupt | test_WSS | test_SP | test_VP |
|---|---|---:|---:|---:|---:|---:|
| H267 EP15+random K=5 | random | 5.9367 | 5.7825 | 6.6898 | 3.6505 | 3.3850 |
| H271 EP13+Sobol K=5 | Sobol | 5.9368 | 5.7797 | 6.6848 | 3.6512 | 3.3864 |
| H274 EP13+anti K=3 | anti | 5.9322 | 5.7763 | 6.6805 | 3.6515 | 3.3846 |
| H275 EP15+anti K=3 ← CURRENT SOTA | anti×EP15 | **5.9243** | **5.7690** | **6.6743** | **3.6427** | **3.3788** |
| **H284 EP15+Sobol-anti K=3 (this)** | Sobol×anti×EP15 | ? | ? | ? | ? | ? |

Plus 2-3 sentences: does Sobol coverage benefit stack with anti-thetic at K=3, or saturate?

**If val < 5.9243 AND test < 5.7690 → immediate SOTA merge**

## Baseline

| Reference | val_abupt | test_abupt | test_WSS | test_VP | test_SP | W&B |
|---|---:|---:|---:|---:|---:|---|
| **H275 EP15+anti-K=3+6-res+mirror ← CURRENT SOTA** | **5.9243** | **5.7690** | **6.6743** | **3.3788** | **3.6427** | 0b4t2bz2 |
| H274 EP13+anti-K=3+6-res+mirror | 5.9322 | 5.7763 | 6.6805 | 3.3846 | 3.6515 | o8oq9r92 |
| H271 EP13+Sobol-K=5+6-res+mirror | 5.9368 | 5.7797 | 6.6848 | 3.3864 | 3.6512 | o9hb87lt |

**Merge gate**: val_abupt < **5.9243** AND test_abupt < **5.7690**

## Critical constraints

- **DDP 8 GPUs** (`--nproc-per-node=8`) for every eval run
- **Read-only**: data/loader.py, data/preload.py, data/split_manifest.json — do not modify
- **No model capacity changes** — eval-time TTA only
- **z-axis tau_z LOCKED at 1.67** (trained value)
- **SENPAI_TIMEOUT_MINUTES=360** (~240 min predicted, comfortable margin)

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT edward:
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["63606nil"],"primary_metric":{"name":"val_abupt_ep15_sobol_anti_K3_stack","value":5.9242},"test_metric":{"name":"test_abupt_ep15_sobol_anti_K3_stack","value":5.7697}}

## Results

**Triple-mechanism stack: Sobol QMC × anti-thetic K=3 × EP15 × 6-res+mirror.**

| Recipe | mechanism | val_abupt | test_abupt | test_WSS | test_SP | test_VP |
|---|---|---:|---:|---:|---:|---:|
| H267 EP15+random K=5 | random | 5.9367 | 5.7825 | 6.6898 | 3.6505 | 3.3850 |
| H271 EP13+Sobol K=5 | Sobol | 5.9368 | 5.7797 | 6.6848 | 3.6512 | 3.3864 |
| H274 EP13+anti K=3 | anti | 5.9322 | 5.7763 | 6.6805 | 3.6515 | 3.3846 |
| H275 EP15+anti K=3 **← CURRENT SOTA** | anti×EP15 | **5.9243** | **5.7690** | **6.6743** | **3.6427** | **3.3788** |
| **H284 EP15+Sobol-anti K=3 (this)** | Sobol×anti×EP15 | **5.9242** | 5.7697 | 6.6745 | 3.6453 | 3.3786 |

### Δ vs H275 SOTA (bp; negative = better)

| metric | H275 | H284 | Δ (bp) | direction |
|---|---:|---:|---:|---|
| val_abupt | 5.92435 | 5.92423 | **−0.012** | ≈ tied (fractionally better) |
| test_abupt | 5.76904 | 5.76969 | **+0.065** | ≈ tied (fractionally worse) |
| test_WSS | 6.67431 | 6.67454 | +0.023 | ≈ tied |
| test_SP | 3.64268 | 3.64530 | +0.262 | slightly worse |
| test_VP | 3.37880 | 3.37858 | −0.022 | ≈ tied |

**Merge gate: val < 5.9243 AND test < 5.7690 → DOES NOT PASS.**
val barely better (−0.012bp), test slightly worse (+0.065bp). The mechanisms do **not stack** at K=3.

### Command

```bash
torchrun --standalone --nproc-per-node=8 eval_tta_h252.py \
  --checkpoint outputs/drivaerml/run-0gjfv45i/checkpoint_ep15.pt \
  --resolutions "32768,49152,65536,81920,98304,131072" \
  --eval-modes "weight_noise_mirror_res_avg" \
  --weight-noise-sigma 5e-4 \
  --weight-noise-passes 3 \
  --antithetic-noise \
  --weight-noise-mode sobol \
  --sobol-proj-dim 1024 \
  --batch-size 2 --num-workers 4 \
  --agent edward \
  --wandb-group "h284-edward-sobol-anti-K3-ep15-stack" \
  --wandb-name "edward/h284-ep15-sobol-anti-K3-stack"
```

- W&B run: [63606nil](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/63606nil)
- Peak GPU memory: **~70 GB / 65 GiB** (DDP×8 H200)
- Wall time: 17 009 s ≈ 283 min (within 360 min budget)
- Eval-only TTA on H185 EP15 checkpoint (`run-0gjfv45i/checkpoint_ep15.pt`); no model/training changes.

### What happened

**Sobol coverage and anti-thetic linear-cancellation saturate at K=3.** Finding UU (Sobol −0.5bp/channel at random K=5, EP13) and Finding VV (anti-thetic linear cancellation at K=3, EP13) each delivered ≈ 0.5bp gains over random K=5. Stacking them on the EP15 / 6-res+mirror base gives **almost exactly zero net movement** vs H275 (val −0.012bp, test +0.065bp — both within run-to-run noise floor < 0.1bp).

Mechanistic read:

1. **Anti-thetic K=3 already kills the dominant variance source.** The +ε / −ε pair zeros the linear Taylor term ∇f(w)·ε exactly, so the residual variance is O(σ²) curvature. Replacing the direction distribution from i.i.d. Gaussian to Sobol-QMC only helps if the quadratic-term variance was direction-coverage-limited — at K=3 with three pair-symmetrised directions, it isn't.
2. **K=3 is too small for QMC to bite.** Sobol's discrepancy advantage scales as O((log K)^d / K) vs i.i.d. O(1/√K). At K=3 in a 1 024-dim Sobol space (projected to 17 M params), the latent points are barely better-spread than three random Gaussians — and after anti-thetic mirroring the effective coverage is 6 antipodal points, leaving the QMC "fill the ball" benefit largely unrealised.
3. **Per-channel deltas are nearly all within noise.** test_WSS (+0.023bp), test_VP (−0.022bp), val (−0.012bp) are all well below the ≤ 0.1bp eval-noise floor seen across the H267 → H275 sweep. Only test_SP shifts measurably (+0.262bp), consistent with the slightly larger test_abupt drift (+0.065bp) being SP-dominated rather than a true mechanism failure.

**Implementation:** ported `_draw_sobol_std_normal` and `_sobol_flat_eps` verbatim from `eval_h271_sobol_tta.py`, added `weight_noise_mode/sobol_proj_dim/sobol_seed/sobol_B_seed/sobol_chunk_size` to `EvalConfig` (auto-wired into CLI via `parse_args`'s `fields(EvalConfig)` loop), and added a `precompute_sobol_eps_per_k` helper that draws one `(K, total_params)` Sobol flat-eps once before evaluation. The anti-thetic pair shares the same eps draw per `k` (linear-term cancellation preserved). Sobol seeds: `sobol_seed=43`, `B_seed=42`, `proj_dim=1024` (matches Finding UU H271).

### Suggested follow-ups

- **Try Sobol-anti at larger K** (K=5 or K=7 anti-thetic pairs, i.e. 10–14 forward passes). At K=3 there's only ~3 distinct directions, well below the regime where QMC discrepancy helps. K≥5 may let Sobol's O((log K)^d/K) advantage show up — but expect ~1.7×–2.3× wall-time per arm.
- **Sobol direction-sharing across resolutions.** Currently each k uses one Sobol-derived ε that is reused across the 6 resolutions × 2 mirrors. An alternative: draw separate Sobol latents per (k, resolution) so the QMC coverage benefits the 12-way res×mirror average rather than just the K dimension. May rescue some of the saturating linear-Taylor benefit if curvature is res-correlated.
- **σ sweep at Sobol-anti K=3.** H275 uses σ=5e-4; Sobol may pair better with a different σ because its eps norms are more uniform than i.i.d. Gaussians' chi-squared norms. Quick σ ∈ {3e-4, 5e-4, 7e-4} sweep at K=1 anti pair (only +2 forward passes per σ) would screen this cheaply.
- **Hold the line on H275.** With Sobol+anti+EP15 stacking saturated and bp-level noise dominating, the next gains likely come from non-eval-time levers (data augmentation, training-time eps schedule, larger EMA window, residual head architecture) rather than further TTA-direction-distribution refinement.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1467 because ADVISOR: Excellent analysis and decomposition. Closing H284 with banked finding.

## Banked: Finding GGG-Sobol-anti-non-stacking-at-K3-EP15

Your mechanistic explanation is correct and important:
- **Anti-thetic K=3 already zeros the linear Taylor term** (∇f·ε cancels exactly between ±ε pairs)
- **Sobol's QMC variance reduction operates on the SAME residual term** as anti-thetic — so they're not orthogonal mechanisms, they're competing for the same variance budget
- At K=3, anti-thetic alone saturates the linear-cancellation gain; Sobol on top adds noise pattern but no further reduction in the dominant variance source
- Net: val −0.012bp / test +0.065bp = effectively zero net movement (both within noise floor)

Combined with **Finding EEE** (anti dominates Sobol at EP15, +4.8bp val / +4.4bp test from H279), this confirms: at K=3+EP15, anti-thetic is the structurally complete mechanism for the linear Taylor source. Sobol contributes nothing complementary.

**Bank Finding GGG**: anti×Sobol compound at K=3 EP15 is non-additive. Mechanisms are not orthogonal — both target the same variance source. Do not pursue further Sobol-on-anti combinations at K≤3.

## What's next for the fleet

The H275 SOTA gate remains 5.9243/5.7690. Current SOTA candidates landing within 1h:
- **fern H288 (8-res mid +40K+57K)**: val 5.9229 (best val signal in fleet!) — test pending
- **tanjiro H285 (K=4 EP15)**: val 5.9235 — test pending

The next viable mechanism axis is **extended-res** (push the resolution envelope). Fern is testing MID densification, frieren is testing LOWER ladder. The UPPER end is untested — that's your H292.

## Decision: Close H284, assign H292 fresh axis

Closing this PR. Banking Finding GGG. Assigning you H292: 8-res UPPER ladder (extended toward higher resolutions). Watch for the assignment PR.

— ADVISOR.
